text/0078-merkleized-metadata.md

```diff
@@ -0,0 +1,410 @@
+
+```rust
+let pruned_types = get_pruned_types();
+
+for ty in pruned_types {
+    if ty.is_primitive_type() {
+        pruned_types.remove(ty);
+    }
+}
+
+pruned_types.sort(|(left, right)|
+    if left.frame_metadata_id() == right.frame_metadata_id() {
+        left.variant_index() < right.variant_index()
+    } else {
+        left.frame_metadata_id() < right.frame_metadata_id()
+    }
+);
+
+fn generate_type_ref(ty, ty_list) -> TypeRef {
+    if ty.is_primitive_type() {
+        TypeRef::primtive_from_ty(ty)
+    }
+
+    TypeRef::from_id(
+        // Determine the id by using the position of the type in the
+        // list of unique frame metadata ids.
+        ty_list.position_by_frame_metadata_id(ty.frame_metadata_id())
+    )
+}
+
+fn replace_all_sub_types_with_type_refs(ty, ty_list) -> Type {
+    for sub_ty in ty.sub_types() {
+        replace_all_sub_types_with_type_refs(sub_ty, ty_list);
+        sub_ty = generate_type_ref(sub_ty, ty_list)
+    }
+
+    ty
+}
+
+let final_ty_list = Vec::new();
+for ty in pruned_types {
+    final_ty_list.push(replace_all_sub_types_with_type_refs(ty, ty_list))
+}
+```
+
+### Building the Merkle Tree Root
+
+A complete binary merkle tree with `blake3` as the hashing function is proposed. For building the merkle tree root, the initial data has to be hashed as a first step. This initial data is referred to as the *leaves* of the merkle tree. The leaves need to be sorted to make the tree root deterministic. The type information is sorted using their unique identifiers and for the `Enumeration`, variants are sort using their `index`. After sorting and hashing all leaves, two leaves have to be combined to one hash. The combination of these of two hashes is referred to as a *node*.
+
+```rust
+let nodes = [];
+while leaves.len() > 1 {
+    let right = leaves.pop_back();
+    let left = leaves.pop_back();
+    nodes.push_back(blake3::hash(scale::encode((left, right))));
+}
+if leaves.len() == 1 {
+    nodes.push_front(leaves.pop_back());
+}
+while nodes.len() > 1 {
+    let right = nodes.pop_front();
+    let left = nodes.pop_front();
+    nodes.push_back(blake3::hash(SCALE::encode((left, right))));
+}
+let merkle_tree_root = if nodes.is_empty() { [0u8; 32] } else { nodes.back() };
```

Comment:
I'd rather simplify the code to make it easier to read. Just setting nodes as leaves and going `ahead!
```suggestion
let nodes = leaves;
while nodes.len() > 1 {
    let right = nodes.pop_back();
    let left = nodes.pop_back();
    nodes.push_front(blake3::hash(scale::encode((left, right))));
}

let merkle_tree_root = if nodes.is_empty() { [0u8; 32] } else { nodes.back() };
```